### PR TITLE
added the TileApi to control real colorful lifx tiles

### PR DESCRIPTION
**Pull request by mabels**
I need to cleanup this PR some sanity checks are missing and 
i was to lazy to build the tests.
I promise that I complete this missing stuff in the next weeks.
This PR is just an reminder to me that i had to completed this!

### CODE REVIEW

**Comment by ristomatti** [comment]
@mabels Based on a quick skim, your PR looks very promising! When finalising the work, please remember to:
- update `README.md`
- follow `.eslintrc` 

Also note that you can use arrow functions even though it's not used much if at all on the existing codebase (the ES2015 rewrite was done quite literally overnight). At least I would prefer arrow functions especially when passing an inline function as a parameter. I can't speak for @Sawtaytoes though who set up this fork though.

**Comment by mabels** [comment]
Hi,

just completed this including tests. There had been 2 other pull request which should applied first.

https://github.com/node-lifx/lifx-lan-client/pull/16
https://github.com/node-lifx/lifx-lan-client/pull/17

Thx in advance

**Comment by ristomatti** [comment]
Massive undertaking. I'll try to review and test this as soon as possible. Sorry if it might take a couple of days! I do have a Tile set to test though.

**Comment by ristomatti** [comment]
@mabels I'm sorry I haven't still had time to review and test this. Just letting you know I haven't forgot this! I just noticed a PR submitted to **node-lifx-lan** a couple of days ago and it'd be cool to have this merged before that. :joy: The problem is though that I'm quite busy today and away from home the rest of the weekend. I'll try to find time today but it doesn't look too good.

**Comment by ristomatti** [comment]
@mabels BTW just noticed there seems to be a conflict due to merging your previous PR #17.

**Comment by ristomatti** [comment]
My plans for today/tomorrow got cancelled. I'll at least start reviewing/testing now.

**Comment by ristomatti** [comment]
@mabels All you need to do to fix the conflict is to drop the already merged commits 3de27da and 
1cca93a and then rebase to upstream master.

**Comment by ristomatti** [comment]
It's actually not a good idea to review the code at it's current state as it contains the already merged changes. @mabels could you update your branch and force push the changes? I'll still test now if I get it to work with my Tiles.

**Comment by ristomatti** [comment]
Sorry a somewhat off-topic note:

> Also note that you can use arrow functions even though it's not used much if at all on the existing codebase (the ES2015 rewrite was done quite literally overnight). At least I would prefer arrow functions especially when passing an inline function as a parameter. I can't speak for @Sawtaytoes though who set up this fork though.

Quoting myself - now when I started to review the changes @mabels has done, I realize there's still two open PR's from me that's aren't included in this repo. When commenting about the arrow functions, I had an odd feeling like I would have done "something" to that already: https://github.com/MariusRumpf/node-lifx/pull/62. This is the *rewrite* I was talking about.

@Sawtaytoes It looks like I forgot PR 62 completely. Would it make sense for me to try to rebase it and do a PR? Or do you prefer working without the use of classes. I've personally been leaning towards more functinal programming approach (mainly at work) but I think using classes would make sense given how the library is written. There'd be a lot less clutter.

File: src/lifx/light.js [hunk 1]
```diff
@@ -71,9 +71,9 @@ Light.prototype.color = function(hue, saturation, brightness, kelvin, duration,
   validate.optionalCallback(callback, 'light color method');
 
   // Convert HSB values to packet format
-  hue = Math.round(hue / constants.HSBK_MAXIMUM_HUE * 65535);
-  saturation = Math.round(saturation / constants.HSBK_MAXIMUM_SATURATION * 65535);
-  brightness = Math.round(brightness / constants.HSBK_MAXIMUM_BRIGHTNESS * 65535);
+  hue = Math.round((hue / constants.HSBK_MAXIMUM_HUE) * 65535);
+  saturation = Math.round((saturation / constants.HSBK_MAXIMUM_SATURATION) * 65535);
+  brightness = Math.round((brightness / constants.HSBK_MAXIMUM_BRIGHTNESS) * 65535);
```
Comment (ristomatti) [inline src/lifx/light.js]: Revert these unrelated changes. It'll be really confusing to see `git blame` say you've last edited the lines but the commit says something about "TileApi".

This would improve the readability a bit I agree, but as this `* 655535` part is repeating so much it would make more sense to refactor it to be done in a function anyway. (But let's still keep refactoring out from this PR).

File: src/lifx/light.js [hunk 2]
```diff
@@ -134,7 +134,7 @@ Light.prototype.colorRgbHex = function(hexString, duration, callback) {
 Light.prototype.maxIR = function(brightness, callback) {
   validate.irBrightness(brightness, 'light setMaxIR method');
 
-  brightness = Math.round(brightness / constants.IR_MAXIMUM_BRIGHTNESS * 65535);
+  brightness = Math.round((brightness / constants.IR_MAXIMUM_BRIGHTNESS) * 65535);
```
Comment (ristomatti) [inline src/lifx/light.js]: Revert these unrelated changes.

File: src/lifx/light.js [hunk 3]
```diff
@@ -510,9 +510,9 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
   validate.optionalCallback(callback, 'color zones method');
 
   // Convert HSB values to packet format
-  hue = Math.round(hue / constants.HSBK_MAXIMUM_HUE * 65535);
-  saturation = Math.round(saturation / constants.HSBK_MAXIMUM_SATURATION * 65535);
-  brightness = Math.round(brightness / constants.HSBK_MAXIMUM_BRIGHTNESS * 65535);
+  hue = Math.round((hue / constants.HSBK_MAXIMUM_HUE) * 65535);
+  saturation = Math.round((saturation / constants.HSBK_MAXIMUM_SATURATION) * 65535);
+  brightness = Math.round((brightness / constants.HSBK_MAXIMUM_BRIGHTNESS) * 65535);
```
Comment (ristomatti) [inline src/lifx/light.js]: Revert these unrelated changes.

File: src/lifx/packets/stateDeviceChain.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
+'use strict';
+
+const Packet = {
+  size:
+    1 +
+    16 * (2 + 2 + 2 + 2 + 4 + 4 + 1 + 1 + 1 + 4 + 4 + 4 + 8 + 8 + 2 + 2 + 4) +
+    1,
```
Comment (ristomatti) [inline src/lifx/packets/stateDeviceChain.js]: The `size` is quite confusing like this. Just use the end result value like on the other packets.

File: src/lifx/packets/stateTileState64.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+'use strict';
+
+const Packet = {
+  size:
+    1 + 1 + 1 + 1 + 1 +
+    64 * (2 + 2 + 2 + 2),
```
Comment (ristomatti) [inline src/lifx/packets/stateTileState64.js]: The `size` is quite confusing like this. Just use the end result value like on the other packets.

File: src/lifx/utils.js [hunk 6]
```diff
@@ -125,20 +186,22 @@ utils.rgbHexStringToObject = function(rgbHexString) {
...
+ * @param {Array} array of numbers
+ * @return {Number} minium of the given array
+ */
 utils.minNumberInArray = function(array) {
-  const sortedCopy = array.slice();
-  sortedCopy.sort(function(a, b) {
-    return a - b;
-  });
-  return sortedCopy[0];
+  return Math.min.apply(undefined, array);
```
Comment (ristomatti) [inline src/lifx/utils.js]: This is a nice cleanup but if the current babel settings allow, use spread instead of `apply`? It's not that important though.

**Review by ristomatti — CHANGES_REQUESTED** [review]
@mabels Added a couple of comments but I'll check the other changes after you've rebased the branch.

**Comment by Sawtaytoes** [comment]
> Sorry a somewhat off-topic note:
> 
> > Also note that you can use arrow functions even though it's not used much if at all on the existing codebase (the ES2015 rewrite was done quite literally overnight). At least I would prefer arrow functions especially when passing an inline function as a parameter. I can't speak for @Sawtaytoes though who set up this fork though.
> 
> Quoting myself - now when I started to review the changes @mabels has done, I realize there's still two open PR's from me that's aren't included in this repo. When commenting about the arrow functions, I had an odd feeling like I would have done "something" to that already: [MariusRumpf/node-lifx#62](https://github.com/MariusRumpf/node-lifx/pull/62). This is the _rewrite_ I was talking about.
> 
> @Sawtaytoes It looks like I forgot PR 62 completely. Would it make sense for me to try to rebase it and do a PR? Or do you prefer working without the use of classes. I've personally been leaning towards more functinal programming approach (mainly at work) but I think using classes would make sense given how the library is written. There'd be a lot less clutter.

Whatever your preference is fine. I think bringing in those old PRs from your other repo would be fantastic!

I also like arrow functions. As long as our Node.js compatibility is fine, then we're good to go; otherwise, we should just bring up the min Node.js version and everything will be easier going forward.

File: src/lifx/packets/stateDeviceChain.js (hunk 4 above)
Comment (Sawtaytoes) [reply src/lifx/packets/stateDeviceChain.js]: I think I get what you're saying @ristomatti. Since those are the UDP groupings from LIFX's docs, I'd instead recommend a comment like this:

```js
size: X // https://docs.lifx.com/some-udp-article
```

Or even:

```js
size: X // This is a grouping of 2, 2, 2, 2, 4, 4, 1, ... etc groupings from LIFX's UDP protocol.
```

Comment (ristomatti) [reply src/lifx/packets/stateDeviceChain.js]: I link to the relevant section of the LAN protocol specification would be sufficient. We don't want to duplicate the documentation here as it occasionally also changes.

Some time ago LIFX has published a machine readable spec document that should make it possible to generate parts of the code. That might be another angle to look into if we'd like to get away with these relatively cryptic packet files. The naming follows the names used in the spec document though so it's quite easy to figure out each part.

**Comment by mabels** [comment]
Node and Arrowfunctions: 

for i in boron carbon dubnium                                            ST 1   master
do
nvm run lts/$i -e '(() => console.log(process.versions.node))()'
done
Running node v6.17.1 (npm v3.10.10)
Running node v8.16.1 (npm v6.4.1)
Running node v10.16.3 (npm v6.11.3)

we should never go to 4.x so on my expirence we are fine.
We could also rewrite arrow functions with babel

File: src/lifx/utils.js (hunk 6 above)
Comment (mabels) [reply src/lifx/utils.js]: mist remerge with the on of the prev PR.
fixed

File: src/lifx/light.js (hunk 2 above)
Comment (mabels) [reply src/lifx/light.js]: I revert this changes, but it drives me crazy.... infact nobody normal knows the order of execution of the operators * /.

File: src/lifx/light.js (hunk 3 above)
Comment (mabels) [reply src/lifx/light.js]: I revert this changes, but it drives me crazy.... infact nobody normal knows the order of execution of the operators * /.

File: src/lifx/light.js (hunk 1 above)
Comment (mabels) [reply src/lifx/light.js]: I revert this changes, but it drives me crazy.... infact nobody normal knows the order of execution of the operators * /.

File: src/lifx/packets/stateDeviceChain.js (hunk 4 above)
Comment (mabels) [reply src/lifx/packets/stateDeviceChain.js]: I simply changed it to the value.
To put such things in the comments is not helpful.
So we lose the source of this arbitrated value, i can live with this.

File: src/lifx/packets/stateTileState64.js (hunk 5 above)
Comment (mabels) [reply src/lifx/packets/stateTileState64.js]: I simply changed it to the value.
To put such things in the comments is not helpful.
So we lose the source of this arbitrated value, i can live with this.

**Comment by mabels** [comment]
> > Sorry a somewhat off-topic note:
> > > Also note that you can use arrow functions even though it's not used much if at all on the existing codebase (the ES2015 rewrite was done quite literally overnight). At least I would prefer arrow functions especially when passing an inline function as a parameter. I can't speak for @Sawtaytoes though who set up this fork though.
> > 
> > 
> > Quoting myself - now when I started to review the changes @mabels has done, I realize there's still two open PR's from me that's aren't included in this repo. When commenting about the arrow functions, I had an odd feeling like I would have done "something" to that already: [MariusRumpf/node-lifx#62](https://github.com/MariusRumpf/node-lifx/pull/62). This is the _rewrite_ I was talking about.
> > @Sawtaytoes It looks like I forgot PR 62 completely. Would it make sense for me to try to rebase it and do a PR? Or do you prefer working without the use of classes. I've personally been leaning towards more functinal programming approach (mainly at work) but I think using classes would make sense given how the library is written. There'd be a lot less clutter.
> 
> Whatever your preference is fine. I think bringing in those old PRs from your other repo would be fantastic!
> 
> I also like arrow functions. As long as our Node.js compatibility is fine, then we're good to go; otherwise, we should just bring up the min Node.js version and everything will be easier going forward.

we should integrate "Arrow functions completly" after this PR is on master.
I offer my help.

File: example/tile.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      brightness: 16384,
+      kelvin: 4096
+    })), () => setBits(light, idx + 1, chain));
+}
+
+client.on('light-new', (light) => {
+  console.log('New light found.');
+  console.log('ID: ' + light.id);
+
+  light.getDeviceChain(function(err, chain) {
```
Comment (ristomatti) [inline example/tile.js:57]: This would now attempt to call `setTileState64` on any light. Add a `const TILE_LABEL = 'MyTile';` and trigger only on a light matching the label.

File: example/tile.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  light.getTileState64(idx, 64, 0, 0, 8, (err) => {
+    if (err) {
+      console.error('getTileState64:', err);
+      return;
+    }
+    getBits(light, idx + 1, chain);
+  });
+}
+
+function setBits(light, idx, chain) {
```
Comment (ristomatti) [inline example/tile.js:35]: As this example has quite a lot going on, it would be great if you could add a comment describing what to expect when this starts running. Maybe include some `console.log`'s printing the values passed to `setTileState64` for example.

File: src/lifx/light.js [hunk 9]
```diff
@@ -529,4 +529,162 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
...
+Light.prototype.setUserPosition = function(tileIndex, userX, userY, reserved, callback) {
+  if (tileIndex === undefined || typeof tileIndex !== 'number') {
+    throw new TypeError('LIFX light setUserPosition tileIndex to be a number');
+  }
+  if (userX === undefined || typeof userX !== 'number') {
+    throw new TypeError('LIFX light setUserPosition userX to be a number');
+  }
+  if (userY === undefined || typeof userY !== 'number') {
+    throw new TypeError('LIFX light setUserPosition userY to be a number');
+  }
```
Comment (ristomatti) [inline src/lifx/light.js]: These validations are now repeated on multiple functions and unnecessarily clutter the function body. Move these to new validation functions in `validator.js` similarly to what's been done for light zone support for example and then replace these with them.

File: src/lifx/light.js [hunk 10]
```diff
@@ -529,4 +529,162 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
...
+ * Requests tile GetTileState64 707
+ *
+ * Get the state of 64 pixels in the tile in a rectangle that has
+ * a starting point and width.
+ * The tileIndex is used to control the starting tile in the chain
+ * and length is used to get the state of that many tiles beginning
+ * from the tileIndex. This will result in a separate response from
+ * each tile.
+ * For the LIFX Tile it really only makes sense to set x and y to
+ * zero, and width to 8.
```
Comment (ristomatti) [inline src/lifx/light.js]: If this comment holds, would it then make sense to move `x`, `y` and `width` as the last parameters and give them a default value using the `function(foo, bar = 'bar')` syntax? If the `babel` preset we have now does not support that, we ought to fix that first. If there's a specific reason for the current order of the parameters, please let me know.

I wouldn't want to rush this change as people might start creating pretty complex animations etc., after which it would be quite rude to change it.

**Review by ristomatti — COMMENTED** [review]
Added few comments. My JS limit for today has maxed out now though. I'll see if I could get this running tomorrow or Sunday.

**Comment by Sawtaytoes** [comment]
@mabels @ristomatti Just your friendly neighborhood ping :).

**Comment by ristomatti** [comment]
> @mabels @ristomatti Just your friendly neighborhood ping :).

I'm still waiting for @mabels. I guess eventually I could try sorting the issues myself but haven't had time to think about this too much unfortunately.

On a related note, I ran across this stuff that's based on the "competing" node-lifx-lan lib that looked pretty cool. Haven't tested it though:
- https://github.com/furey/lifx-tile-effects-framework
- https://github.com/furey/lifx-tile-halloween-effects

File: example/tile.js (hunk 7 above)
Comment (mabels) [reply example/tile.js:57]: done

File: example/tile.js (hunk 8 above)
Comment (mabels) [reply example/tile.js:35]: done

File: src/lifx/light.js (hunk 9 above)
Comment (mabels) [reply src/lifx/light.js]: done

File: src/lifx/light.js (hunk 10 above)
Comment (mabels) [reply src/lifx/light.js]: i change the api of set/get TileState64 to an option prop style
where the defaults are applied safely.

**Review by Sawtaytoes — APPROVED** [review]

**Comment by ristomatti** [comment]
Thanks @mabels. I will check and test this later today (= within the next 12 hours). If I won't, you're welcome to merge this @Sawtaytoes.

**Comment by Sawtaytoes** [comment]
@ristomatti Merging in 1 hour.

**Comment by ristomatti** [comment]
@Sawtaytoes Thanks for the heads-up (although I saw it only just now). Sadly I was blocked by a force majeure.

Would it be possible for me to email or PM you on this? You can find my email from my commit messages. Your location indicates you're 8 hours behind but I'm a night owl so there should be plenty of hours reach me.

**Comment by ristomatti** [comment]
Impossible things can happen - testing now.

**Comment by ristomatti** [comment]
@mabels Did you test this after your changes? I can't get this to work. There's multiple issues. When running `example/tile.js` I got this:

```
$ node tile.js 
Started LIFX listening on 0.0.0.0:49419

New light found.
ID: d073d5394495
/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657
  validate.isUint8(tileIndex, 'setTileState64', 'tileIndex');
           ^

TypeError: validate.isUint8 is not a function
    at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657:12)
    at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
    at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
    at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
```

After replacing renaming 3 calls to `validate.isUint8` with `validate.isUInt8` it crashed like so:

```
$ node tile.js 
Started LIFX listening on 0.0.0.0:43892

New light found.
ID: d073d5394495
/home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143
    throw new TypeError('LIFX util buildColorsHsbk expects colors to be an array');
    ^

TypeError: LIFX util buildColorsHsbk expects colors to be an array
    at Object.utils.buildColorsHsbk (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143:11)
    at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:671:27)
    at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
    at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
    at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
```

File: example/tile.js (hunk 7 above)
Comment (ristomatti) [reply example/tile.js:57]: This does not work as light label has not resolved at this point. It was partly my bad as I didn't remember this while reviewing this the first time. It would improve the example a lot if a label could be used but for this purpose it would be enough to just rename `TILE_LABEL` TO `TILE_LIGHT_ID` for example.

**Comment by ristomatti** [comment]
@Sawtaytoes Sorry but this is not ready yet. I don't think we should rush this in as there might be other issues also. For instance `README.md` would need to be updated.

File: src/lifx/light.js [hunk 11]
```diff
@@ -529,4 +529,159 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
...
+  }, sqnNumber);
+};
+
+/**
+ * Sets Tile Position
+ * @example light.setUserPosition(0, 12, 13, 0, () => {})
+ * @param {Number} tileIndex	unsigned 8-bit integer
+ * @param {Number} userX	32-bit float
+ * @param {Number} userY	32-bit float
+ * @param {Number} reserved	16-bit integer
```
Comment (ristomatti) [inline src/lifx/light.js:556]: What is this `reserved` value and how would a user of this library know what to put there? This applies to all functions.

File: src/lifx/light.js [hunk 12]
```diff
@@ -529,4 +529,159 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
...
+ * @param {Number} options.width	unsigned 8-bit integer (default 8)
+ * @param {Number} options.reserved	unsigned 8-bit integer
+ * @param {Function} callback a function to accept the data
+ */
+Light.prototype.getTileState64 = function(tileIndex, options, callback) {
+  validate.isUint8(tileIndex, 'getTileState64', 'tileIndex');
+  if (typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
```
Comment (ristomatti) [inline src/lifx/light.js:624]: What's the purpose of the above block?

File: src/lifx/light.js [hunk 13]
```diff
@@ -529,4 +529,159 @@ Light.prototype.colorZones = function(startIndex, endIndex, hue, saturation, bri
...
+ * @param {Number} options.duration	unsigned 32-bit integer (default 0)
+ * @param {Number} options.reserved	unsigned 8-bit integer
+ * @param {Function} [callback] called when light did receive message
+ */
+Light.prototype.setTileState64 = function(tileIndex, colors, options, callback) {
+  validate.isUint8(tileIndex, 'setTileState64', 'tileIndex');
+  if (typeof options === 'function') {
+    callback = options;
+    options = {};
+  }
```
Comment (ristomatti) [inline src/lifx/light.js:668]: What's the purpose of the above block? Also note that `options` is assigned to `{}` but `options` is not used after this.

**Review by ristomatti — CHANGES_REQUESTED** [review]

**Comment by mabels** [comment]
with the unused options that is my fault.

I will fix it and build a test that it not happend again

meno

File: src/lifx/light.js (hunk 13 above)
Comment (mabels) [reply src/lifx/light.js:668]: this was a typo

File: src/lifx/light.js (hunk 12 above)
Comment (mabels) [reply src/lifx/light.js:624]: the options are optional  that you could pass on parameter options the callback

**Comment by mabels** [comment]
> @mabels Did you test this after your changes? I can't get this to work. There's multiple issues. When running `example/tile.js` I got this:
> 
> ```
> $ node tile.js 
> Started LIFX listening on 0.0.0.0:49419
> 
> New light found.
> ID: d073d5394495
> /home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657
>   validate.isUint8(tileIndex, 'setTileState64', 'tileIndex');
>            ^
> 
> TypeError: validate.isUint8 is not a function
>     at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657:12)
>     at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
>     at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
>     at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
> ```
> 
> After replacing renaming 3 calls to `validate.isUint8` with `validate.isUInt8` it crashed like so:
> 
> ```
> $ node tile.js 
> Started LIFX listening on 0.0.0.0:43892
> 
> New light found.
> ID: d073d5394495
> /home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143
>     throw new TypeError('LIFX util buildColorsHsbk expects colors to be an array');
>     ^
> 
> TypeError: LIFX util buildColorsHsbk expects colors to be an array
>     at Object.utils.buildColorsHsbk (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143:11)
>     at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:671:27)
>     at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
>     at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
>     at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
> ```

I could not test this change off course i had no lifx on hand.
I will try this if i come back home in the next week.

**Comment by mabels** [comment]
> > @mabels Did you test this after your changes? I can't get this to work. There's multiple issues. When running `example/tile.js` I got this:
> > ```
> > $ node tile.js 
> > Started LIFX listening on 0.0.0.0:49419
> > 
> > New light found.
> > ID: d073d5394495
> > /home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657
> >   validate.isUint8(tileIndex, 'setTileState64', 'tileIndex');
> >            ^
> > 
> > TypeError: validate.isUint8 is not a function
> >     at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:657:12)
> >     at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
> >     at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
> >     at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
> > ```
> > 
> > 
> > After replacing renaming 3 calls to `validate.isUint8` with `validate.isUInt8` it crashed like so:
> > ```
> > $ node tile.js 
> > Started LIFX listening on 0.0.0.0:43892
> > 
> > New light found.
> > ID: d073d5394495
> > /home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143
> >     throw new TypeError('LIFX util buildColorsHsbk expects colors to be an array');
> >     ^
> > 
> > TypeError: LIFX util buildColorsHsbk expects colors to be an array
> >     at Object.utils.buildColorsHsbk (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/utils.js:143:11)
> >     at Light.setTileState64 (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:671:27)
> >     at setBits (/home/ristomatti/Projects/lifx-lan-client/example/tile.js:41:9)
> >     at /home/ristomatti/Projects/lifx-lan-client/example/tile.js:61:5
> >     at Client.<anonymous> (/home/ristomatti/Projects/lifx-lan-client/lib/lifx/light.js:534:12)
> > ```
> 
> I could not test this change off course i had no lifx on hand.
> I will try this if i come back home in the next week.

**Comment by mabels** [comment]
I need to build a test for the high order function. I'm sorry for this chain of failures.
